<a href="https://colab.research.google.com/github/ikwak2/StatBook/blob/main/ch8/8_10_R_prac.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 8.10 R 연습

## <예 8-2>의 단일집단의 t검정 과정
### 자료 입력

In [1]:
x1<-c(10.8,12.2,9.5,10.3,8.6,8.3,9.5,10,11.5,10.1)
mean(x1)
sd(x1)
length(x1) #x1에 속한 원소의 개수 (=n)

[1] 10.08

[1] 1.207201

[1] 10

x1에 검정에 필요한 기초 자료를 저장한다. 앞의 기본 함수들을 이용하여 요약통계량을 구한다. length(x1)은 x1 자료의 길이로 표본 수 n에 해당한다.

### 가설검정

In [2]:
t.test(x1,mu=10)


	One Sample t-test

data:  x1
t = 0.20956, df = 9, p-value = 0.8387
alternative hypothesis: true mean is not equal to 10
95 percent confidence interval:
  9.216421 10.943579
sample estimates:
mean of x 
    10.08 


앞의 결과는 $H_0: \mu=10$을 검정하기 위한 명령문과 결과이다. R에서 기본으로 제공되는 명령어 t.test를 이용하며, 자료, 귀무가설 하의 , 검정의 종류를 순서대로 입력하여 사용한다. 위 그림의 명령문 t.test(x1,mu=10)에서 x1은 자료, mu=10은 귀무가설 하의 를 10으로 하겠다는 것을 의미한다. 대립가설의 종류를 따로 입력하지 않으면 기본값으로 양측검정을 시행하게 된다. 명령문을 입력하면 단일집단 t검정의 결과가 출력된다. data는 사용한 자료, t는 검정통계량의 값(t값), df는 자유도, p-value는 유의확률(p값)을 나타낸다.  
$H_0: \mu=10$에 대한 대립가설을 따로 지정하지 않았으므로 $H_1: \mu\ne10$에 대한 검정을 시행했다. 

### 검정 결과
p-value가 0.8387로 유의수준 5% 하에서 귀무가설을 기각하지 못한다. 즉, 위의 표본관측 결과에 의할 때 표본(x1)의 평균이 10이 아니라고 할 수 없다. 결과창에서 $\mu$의 95% 신뢰구간과 표본평균도 함께 출력된다. 

## 두 모집단 t검정 과정(두 모집단의 분산이 동일하지 않다고 전제할 수 있는 경우)
### 자료 입력

In [3]:
set.seed(1)
A<-round(rnorm(100,70,10))
B<-round(rnorm(100,80,15))
A
B 

[1] 64 72 62 86 73 62 75 77 76 67 85 74 64 48 81 70 70 79 78 76 79 78 71 50 76
 [26] 69 68 55 65 74 84 69 74 69 56 66 66 69 81 78 68 67 77 76 63 63 74 78 69 79
 [51] 74 64 73 59 84 90 66 60 76 69 94 70 77 70 63 72 52 85 72 92 75 63 76 61 57
 [76] 73 66 70 71 64 64 69 82 55 76 73 81 67 74 73 65 82 82 77 86 76 57 64 58 65

[1]  71  81  66  82  70 107  91  94  86 105  70  73 101  70  77  74  75  76
 [19]  87  77  72 100  77  77  78  91  79  79  70  75  81  71  88  57  85  57
 [37]  75  72  70  79  51  98  55  73  63  69 111  80  61  55  87  80  75  66
 [55]  58  64  95  71  59 108  86  76  96  93  71 113  76  59  78  83 115  82
 [73]  87  79  75  79  92 111  95  98  62  95  83  58  88  78 102  69  74  66
 [91]  77  86  69  92  62  64 102  65  86  74

두 모집단의 모분산이 동일하지 않을 경우를 가정한 t검정의 예제를 위해 정규분포를 따르는 크기 100의 적당한 자료벡터를 2개 생성한다. <자료 A>는 평균 70, 표준편차 10인 정규분포에서 랜덤하게 추출한 100개의 표본이고, <자료 B>는 평균 80, 표준편차 15인 정규분포에서 랜덤하게 추출한 100개의 표본이다. 여기서 set.seed(1)은 랜덤하게 생성되는 난수를 고정하기 위해 사용되었고, round는 소수점을 반올림하여 자연수를 만들어주는 명령어이다. t검정을 위해서는 정규성 검정이 선행되어야 하지만 이 예제에서는 정규분포로부터 표본을 만들었으므로 생략한다.

### 가설검정

In [4]:
t.test(A,B,var.equal=F)


	Welch Two Sample t-test

data:  A and B
t = -4.8743, df = 166.28, p-value = 2.532e-06
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -11.619745  -4.920255
sample estimates:
mean of x mean of y 
    71.14     79.41 


위 결과는 $H_0:\mu_A = \mu_B$ 를 검정하기 위한 명령문과 결과이다. 단일집단 t검정과 동일하게 R에서 기본으로 제공되는 명령어 t.test를 이용한다. 이 경우 검정을 위한 두 표본자료, 등분산 가정 여부, 검정의 종류를 순서대로 입력하여 사용한다. 위 그림의 명령문 t.test(A,B,var.equal=F)에서 A와 B는 표본자료, var.equal=F는 등분산을 가정하지 않겠다는 것을 의미한다. 여기서 F는 FALSE로도 입력할 수 있다. 등분산을 가정하는 경우는 T 혹은 TRUE로 입력한다. 대립가설의 종류를 따로 입력하지 않으면 기본값으로 양측검정을 시행하게 된다. 명령문을 입력하면 두 모집단 t검정의 결과가 출력된다.  
$H_0:\mu_A = \mu_B$에 대한 대립가설을 따로 지정하지 않았으므로 $H_1:\mu_A \ne \mu_B$에 대한 검정을 시행했다. 

### 검정 결과

p-value를 볼 때, 유의수준 5% 하에서 귀무가설을 기각한다. 즉, 위의 표본관측 결과에 의할 때 유의수준 5% 하에서 표본 A와 표본 B의 모집단 평균이 같다고 할 수 없다.

## 두 모집단 t검정 과정(두 모집단의 분산이 동일하다고 전제할 수 있는 경우)
### 자료 입력

In [5]:
set.seed(1)
A<-round(rnorm(100,70,10))
B<-round(rnorm(100,80,10))
A
B

[1] 64 72 62 86 73 62 75 77 76 67 85 74 64 48 81 70 70 79 78 76 79 78 71 50 76
 [26] 69 68 55 65 74 84 69 74 69 56 66 66 69 81 78 68 67 77 76 63 63 74 78 69 79
 [51] 74 64 73 59 84 90 66 60 76 69 94 70 77 70 63 72 52 85 72 92 75 63 76 61 57
 [76] 73 66 70 71 64 64 69 82 55 76 73 81 67 74 73 65 82 82 77 86 76 57 64 58 65

[1]  74  80  71  82  73  98  87  89  84  97  74  75  94  73  78  76  77  77
 [19]  85  78  75  93  78  78  79  87  79  80  73  77  81  74  85  65  83  65
 [37]  77  75  73  79  61  92  63  75  69  72 101  80  67  64  85  80  77  71
 [55]  65  69  90  74  66  99  84  78  91  89  74 102  77  66  79  82 103  81
 [73]  85  79  77  80  88 101  90  92  68  90  82  65  85  78  95  72  76  71
 [91]  78  84  73  88  68  70  94  70  84  76

두 모집단의 모분산이 동일한 경우를 가정한 t검정의 예제를 위해 정규분포를 따르는 크기 100의 적당한 자료벡터를 2개 생성한다. <자료 A>는 평균 70, 표준편차 10인 정규분포에서 랜덤하게 추출한 100개의 표본이고, <자료 B>는 평균 80, 표준편차 10인 정규분포에서 랜덤하게 추출한 100개의 표본이다. 마찬가지로 정규성 검정은 생략한다.

### 가설검정

In [6]:
t.test(A,B,var.equal=T)


	Two Sample t-test

data:  A and B
t = -6.4574, df = 198, p-value = 8.056e-10
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -11.08275  -5.89725
sample estimates:
mean of x mean of y 
    71.14     79.63 


위 그림은 $H_0:\mu_A = \mu_B$를 검정하기 위한 명령문과 결과이다. 명령문은 앞의 경우와 같고 등분산을 가정하고 있으므로 var.equal=T를 입력한다. 여기서 T는 TRUE로도 입력할 수 있다. 대립가설의 종류를 따로 입력하지 않으면 기본값으로 양측검정을 시행하게 된다. 명령문을 입력하면 두 모집단의 t검정 결과가 출력된다.  
$H_0:\mu_A = \mu_B$에 대한 대립가설을 따로 지정하지 않았으므로 $H_1:\mu_A \ne \mu_B$에 대한 검정을 시행했다.

3) 검정 결과
p-value로 볼 때 유의수준 5% 하에서 귀무가설을 기각한다. 즉, 위의 표본관측 결과에 의할 때 유의수준 5% 하에서 표본 A와 표본 B의 모집단 평균이 같다고 할 수 없다.

## <예 8-7>의 짝진표본 t검정 과정
### 자료 입력

In [7]:
A<-c(10.6,9.4,12.3,9.7,8.3)
B<-c(10.2,9.8,11.8,9.1,8.8)
A
B

[1] 10.6  9.4 12.3  9.7  8.3

[1] 10.2  9.8 11.8  9.1  8.8

A와 B에 검정에 필요한 기초 자료를 입력한다. 짝진표본 t검정의 경우 두 자료벡터 A, B에 입력되는 자료의 순서가 동일해야 한다. <예 8-7>에서 10.6, 10.2는 동일한 자동차에서 실험된 타이어의 마모율이므로, 각각 첫 번째 열에 입력한다. 이처럼 순서가 같은 열에 있는 자료는 서로 짝진 관계에 있어야 하고, 두 자료벡터의 길이는 같아야 한다.

### 가설검정

In [9]:
t.test(A,B,paired=T,alt = "two.sided")


	Paired t-test

data:  A and B
t = 0.50983, df = 4, p-value = 0.637
alternative hypothesis: true mean difference is not equal to 0
95 percent confidence interval:
 -0.5334976  0.7734976
sample estimates:
mean difference 
           0.12 


앞의 그림은 $H_0:\mu_A = \mu_B$ 를 검정하기 위한 명령문과 결과이다. 명령문에서 A, B는 검정에 사용되는 자료, paired=T는 짝진표본 t검정을 사용하겠다는 의미이다. 여기서 T는 TRUE로도 입력할 수 있다. 입력하지 않으면 기본값으로 두 모집단 t검정을 실시하게 된다. alt=’two.sided’에서 대립가설의 종류를 양측검정으로 지정했다. ‘  ’안에 greater, less를 입력함으로써 대립가설의 종류를 바꿀 수 있다. alt=‘대립가설’, alternative=‘대립가설’과 같이 입력한다.
$H_0:\mu_A = \mu_B$, $H_1:\mu_A \ne \mu_B$에 대한 검정을 시행했다.


3) 검정 결과
p-value가 0.637로 유의수준 5% 하에서 귀무가설을 기각할 수 없다. 즉, 5% 유의수준 하에서 두 타이어의 성능은 동일하다고 볼 수 있다.

※ 대립가설의 종류 변경

In [10]:
t.test(A,B,paired=T,alt = "greater")


	Paired t-test

data:  A and B
t = 0.50983, df = 4, p-value = 0.3185
alternative hypothesis: true mean difference is greater than 0
95 percent confidence interval:
 -0.3817771        Inf
sample estimates:
mean difference 
           0.12 


위 그림은 대립가설의 종류를 ‘greater’로 지정했을 경우의 검정 결과이다. 이 경우 $H_0:\mu_A = \mu_B$, $H_1:\mu_A > \mu_B$ 에 대한 검정을 시행했다.

## <예 8-9>의 모비율 P에 대한 검정과정
### 가설검정

In [11]:
prop.test(10,300,p=0.05,alt="less",correct=F)


	1-sample proportions test without continuity correction

data:  10 out of 300, null probability 0.05
X-squared = 1.7544, df = 1, p-value = 0.09266
alternative hypothesis: true p is less than 0.05
95 percent confidence interval:
 0.00000000 0.05497991
sample estimates:
         p 
0.03333333 


R에서 기본으로 제공하는 모비율 검정 명령어 prop.test를 이용한다. prop. test에서는 따로 자료를 입력하지 않는다. 명령문에서 괄호 안은 성공의 횟수, 총 시행횟수, 귀무가설 하의 모비율, 대립가설의 종류 순이다. ‘correct=F’는 연속성 보정에 관한 부분이며, 일반적으로 ‘correct=F’로 사용한다. 입력하지 않거나, T 혹은 TRUE를 입력하면 연속성 보정된 검정을 시행한다.  
$H_0:P = 0.05$, $H_1:P < 0.05$에 대한 검정을 시행했다.

### 결과 해석

p-value가 0.09266으로 유의수준 5% 하에서 귀무가설을 기각할 수 없다.  즉, 주어진 표본관측 결과에 의할 때 이 회사 제품의 불량률이 5%보다 적다고 할 수 없다.

## <예 8-14>의 두 모분산의 동일성 검정
### 자료 입력

In [12]:
set.seed(1)
A<-round(rnorm(100,70,10))
B<-round(rnorm(100,80,15))
A
B

[1] 64 72 62 86 73 62 75 77 76 67 85 74 64 48 81 70 70 79 78 76 79 78 71 50 76
 [26] 69 68 55 65 74 84 69 74 69 56 66 66 69 81 78 68 67 77 76 63 63 74 78 69 79
 [51] 74 64 73 59 84 90 66 60 76 69 94 70 77 70 63 72 52 85 72 92 75 63 76 61 57
 [76] 73 66 70 71 64 64 69 82 55 76 73 81 67 74 73 65 82 82 77 86 76 57 64 58 65

[1]  71  81  66  82  70 107  91  94  86 105  70  73 101  70  77  74  75  76
 [19]  87  77  72 100  77  77  78  91  79  79  70  75  81  71  88  57  85  57
 [37]  75  72  70  79  51  98  55  73  63  69 111  80  61  55  87  80  75  66
 [55]  58  64  95  71  59 108  86  76  96  93  71 113  76  59  78  83 115  82
 [73]  87  79  75  79  92 111  95  98  62  95  83  58  88  78 102  69  74  66
 [91]  77  86  69  92  62  64 102  65  86  74


두 모집단 t검정 과정에서 사용했던 자료를 이용하여 두 모분산의 동일성 검정을 시행한다. 이 경우 A와 B는 모분산이 다른 정규분포에서 랜덤하게 생성한 자료이므로 어느 정도 결과를 예상할 수 있다.

### 가설검정 

In [13]:
var.test(A,B)


	F test to compare two variances

data:  A and B
F = 0.39205, num df = 99, denom df = 99, p-value = 4.973e-06
alternative hypothesis: true ratio of variances is not equal to 1
95 percent confidence interval:
 0.2637849 0.5826720
sample estimates:
ratio of variances 
          0.392046 


R에서 기본으로 제공하는 명령어 var.test를 이용한다. 명령문의 괄호 안에는 비교대상인 두 표본을 입력한다.  
$H_0:\sigma^2_A = \sigma^2_B$, $H_1:\sigma^2_A \ne \sigma^2_B$에 대한 검정을 시행했다.

### 결과 해석
p-value가 4.973e-06 이므로, 유의수준 5% 하에서 귀무가설을 기각한다. 즉, 두 표본 A, B의 모분산이 동일하다고 할 수 없다.